kroxylicious/src/main/java/io/kroxylicious/proxy/filter/FilterChainContributor.java

```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.kroxylicious.proxy.filter;
+
+import java.util.Map;
+
+/**
+ * Implementations contribute one or more filters to the filter chain.
+ */
+public interface FilterChainContributor {
```

Comment:
The idea of contributors is based on the insight (well, hopefully, anyways), that features are of a semantic nature, say "multi-tenancy", which will require to have several filters at different locations of the filter chain in place. So a contributor can contribute filters at different ordinal values. It's all a bit muddy still, but essentially a filter early in the chain would have a priority of 1, and a filter right before messages are sent upstream would have a priority of 1000 (and the reverse for responses). Conflict resolution still is an open question, but then again, baby steps...

Reply:
I'm not certain I see that a single feature (like multi-tenancy) would require multiple filters. Remember that a filter implementation can multiply inherit the protocol filter interfaces, so a single instance could do things like renaming topics, groups ids and transactional ids. If you have multiple filters making up a feature then you might also end up with a requirement for them to communicate state between themselves. Did you have a specific example shows that multiple filters might really be needed to implement a use case?

Reply:
Fair question, perhaps I'm for once erring on the side of offering too much flexibility just yet. I felt the topic rename feature _may_ need it, also perhaps it might be nice to split up the implementation of one "feature" across multiple separate filter classes (good point though on the need to exchange state). Or there could be one contributor which bulk-provides multiple features? Long story short: It's not strictly needed _right now_, and we can remove it, if you feel strongly about it.